### ISSUE TITLE
pmlogger.service missing from .deb after ./Makepkgs

### ISSUE THREAD

**Issue by rvernica** [performancecopilot/pcp#16]
I compiled version `3.10.4` from Git with `./Makepkgs`. After installing the `.deb` files, it seems that the `/lib/systemd/system/pmlogger.service` file is missing. Here is the error message that I get when installing the `.deb` files:
```
$ sudo dpkg -i pcp_3.10.4_armhf.deb libpcp-gui2_3.10.4_armhf.deb libpcp-mmv1_3.10.4_armhf.deb libpcp-pmda3_3.10.4_armhf.deb libpcp-trace2_3.10.4_armhf.deb libpcp3_3.10.4_armhf.deb libpcp-pmda-perl_3.10.4_armhf.deb python-pcp_3.10.4_armhf.deb pcp-conf_3.10.4_armhf.deb libpcp-import1_3.10.4_armhf.deb 
Selecting previously unselected package pcp.
(Reading database ... 61382 files and directories currently installed.)
Unpacking pcp (from pcp_3.10.4_armhf.deb) ...
Selecting previously unselected package libpcp-gui2.
Unpacking libpcp-gui2 (from libpcp-gui2_3.10.4_armhf.deb) ...
Selecting previously unselected package libpcp-mmv1.
Unpacking libpcp-mmv1 (from libpcp-mmv1_3.10.4_armhf.deb) ...
Selecting previously unselected package libpcp-pmda3.
Unpacking libpcp-pmda3 (from libpcp-pmda3_3.10.4_armhf.deb) ...
Selecting previously unselected package libpcp-trace2.
Unpacking libpcp-trace2 (from libpcp-trace2_3.10.4_armhf.deb) ...
Selecting previously unselected package libpcp3.
Unpacking libpcp3 (from libpcp3_3.10.4_armhf.deb) ...
Selecting previously unselected package libpcp-pmda-perl.
Unpacking libpcp-pmda-perl (from libpcp-pmda-perl_3.10.4_armhf.deb) ...
Selecting previously unselected package python-pcp.
Unpacking python-pcp (from python-pcp_3.10.4_armhf.deb) ...
Selecting previously unselected package pcp-conf.
Unpacking pcp-conf (from pcp-conf_3.10.4_armhf.deb) ...
Selecting previously unselected package libpcp-import1.
Unpacking libpcp-import1 (from libpcp-import1_3.10.4_armhf.deb) ...
Setting up pcp-conf (3.10.4) ...
Setting up libpcp3 (3.10.4) ...
Setting up libpcp-import1 (3.10.4) ...
Setting up libpcp-gui2 (3.10.4) ...
Setting up libpcp-mmv1 (3.10.4) ...
Setting up libpcp-pmda3 (3.10.4) ...
Setting up libpcp-trace2 (3.10.4) ...
Setting up libpcp-pmda-perl (3.10.4) ...
Setting up python-pcp (3.10.4) ...
Processing triggers for man-db ...
Setting up pcp (3.10.4) ...
Rebuilding PMNS ... 
Starting pmcd ... 
Failed to issue method call: No such file or directory
/etc/init.d/pmlogger: Warning: Performance Co-Pilot archive logger(s) not permanently enabled.
    To enable pmlogger, run the following as root:
    # /bin/systemctl enable pmlogger.service
Starting pmlogger ... 
Processing triggers for desktop-file-utils ...
```
Here are all the `pmlogger` files that show up in the system:
```
$ locate pmlogger
/etc/cron.d/pcp-pmlogger
/etc/init.d/pmlogger
/etc/pcp/pmlogger
/etc/pcp/pmlogger/control
/etc/rc0.d/K01pmlogger
/etc/rc1.d/K01pmlogger
/etc/rc2.d/S03pmlogger
/etc/rc3.d/S03pmlogger
/etc/rc4.d/S03pmlogger
/etc/rc5.d/S03pmlogger
/etc/rc6.d/K01pmlogger
/usr/bin/pmlogger
/usr/lib/pcp/bin/pmlogger
/usr/lib/pcp/bin/pmlogger_check
/usr/lib/pcp/bin/pmlogger_daily
/usr/lib/pcp/bin/pmlogger_merge
/usr/share/man/man1/pmlogger.1.gz
/usr/share/man/man1/pmlogger_check.1.gz
/usr/share/man/man1/pmlogger_daily.1.gz
/usr/share/man/man1/pmlogger_merge.1.gz
/usr/share/pcp/demos/pmclient/pmlogger.config
/var/lib/pcp/config/pmlogger
/var/lib/pcp/config/pmlogger/config.default
/var/lib/pcp/config/pmlogger/config.pmstat
/var/lib/pcp/tmp/pmlogger
/var/lib/pcp/tmp/pmlogger/24522
/var/lib/pcp/tmp/pmlogger/primary
/var/log/pcp/pmlogger
/var/log/pcp/pmlogger/beaglebone
/var/log/pcp/pmlogger/beaglebone/20150421.21.03.0
/var/log/pcp/pmlogger/beaglebone/20150421.21.03.index
/var/log/pcp/pmlogger/beaglebone/20150421.21.03.meta
/var/log/pcp/pmlogger/beaglebone/Latest
/var/log/pcp/pmlogger/beaglebone/pmlogger.log
```
Here are all the `pmlogger` related files contained in all the `.deb` files generated:
```
 for f in *.deb; do dpkg -c $f; done | grep pmlogger
-rwxr-xr-x root/root     21911 2015-04-21 19:03 ./usr/lib/pcp/bin/pmlogger_check
-rwxr-xr-x root/root      5929 2015-04-21 19:03 ./usr/lib/pcp/bin/pmlogger_merge
-rwxr-xr-x root/root     23362 2015-04-21 19:03 ./usr/lib/pcp/bin/pmlogger_daily
-rwxr-xr-x root/root     80096 2015-04-21 19:09 ./usr/bin/pmlogger
-rw-r--r-- root/root       306 2015-04-21 19:02 ./usr/share/pcp/demos/pmclient/pmlogger.config
-rw-r--r-- root/root      7208 2015-04-21 19:05 ./usr/share/man/man1/pmlogger.1.gz
-rw-r--r-- root/root      5092 2015-04-21 19:05 ./usr/share/man/man1/pmlogger_check.1.gz
drwxrwxr-x root/root         0 2015-04-21 19:03 ./var/lib/pcp/tmp/pmlogger/
drwxr-xr-x root/root         0 2015-04-21 19:03 ./var/lib/pcp/config/pmlogger/
-rw-r--r-- root/root       560 2015-04-21 19:03 ./var/lib/pcp/config/pmlogger/config.pmstat
drwxr-xr-x root/root         0 2015-04-21 19:03 ./var/log/pcp/pmlogger/
-rwxr-xr-x root/root      7642 2015-04-21 19:03 ./etc/init.d/pmlogger
drwxr-xr-x root/root         0 2015-04-21 19:03 ./etc/pcp/pmlogger/
-rw-r--r-- root/root      1700 2015-04-21 19:03 ./etc/pcp/pmlogger/control
-rw-r--r-- root/root       355 2015-04-21 19:03 ./etc/cron.d/pcp-pmlogger
lrwxrwxrwx root/root         0 2015-04-21 19:08 ./usr/lib/pcp/bin/pmlogger -> ../../../bin/pmlogger
lrwxrwxrwx root/root         0 2015-04-21 19:08 ./usr/share/man/man1/pmlogger_daily.1.gz -> pmlogger_check.1.gz
lrwxrwxrwx root/root         0 2015-04-21 19:08 ./usr/share/man/man1/pmlogger_merge.1.gz -> pmlogger_check.1.gz
-rw-r--r-- root/root     11342 2015-04-21 19:06 ./usr/share/doc/pcp-doc/html/lab.pmlogger.html
-rw-r--r-- root/root         0 2015-04-21 19:03 ./etc/pcp/pmmgr/pmlogger
```
Finally, the file seems to be present in the `build` directory:
```
build/deb/pcp-3.10.4/src/pmlogger/pmlogger.service
build/deb/pcp-3.10.4/src/pmlogger/pmlogger.service.in
```
So, somehow this file eludes the `.deb` files. Thanks!

**Comment by kmcdonell** [comment]
Hmm, thanks for this ... I think.
What sort of Debian-based system is this?
This will happen if the configure magic determines that ENABLE_SYSTEMD is false ... can you please post the results of:
$ grep ENABLE_SYSTEMD /usr/include/pcp/builddefs
$ grep PCP_SYSTEMDUNIT_DIR /etc/pcp.conf

**Comment by rvernica** [comment]
`grep` output:
```
$ grep ENABLE_SYSTEMD /usr/include/pcp/builddefs
ENABLE_SYSTEMD = true
$ grep PCP_SYSTEMDUNIT_DIR /etc/pcp.conf
PCP_SYSTEMDUNIT_DIR=/lib/systemd/system
```
Debian info:
```
$ cat /etc/issue
Debian GNU/Linux 7
$ uname -a
Linux beaglebone 3.8.13-bone50 #1 SMP Tue May 13 13:24:52 UTC 2014 armv7l GNU/Linux
```
I just noticed that other `.service` files are missing as well. I can start `pmcd.service` but I can't `enable` it:
```
$ sudo systemctl start pmcd.service
$ sudo systemctl status pmcd.service
pmcd.service - LSB: Control pmcd (the collection daemon for PCP)
	  Loaded: loaded (/etc/init.d/pmcd)
	  Active: active (running) since Tue, 21 Apr 2015 23:52:15 +0000; 5s ago
	 Process: 6570 ExecStart=/etc/init.d/pmcd start (code=exited, status=0/SUCCESS)
	  CGroup: name=systemd:/system/pmcd.service
		  ├ 6693 /usr/lib/pcp/bin/pmcd
		  ├ 6697 /var/lib/pcp/pmdas/root/pmdaroot
		  ├ 6698 /var/lib/pcp/pmdas/proc/pmdaproc -d 3
		  ├ 6699 /var/lib/pcp/pmdas/xfs/pmdaxfs -d 11
		  └ 6700 /var/lib/pcp/pmdas/linux/pmdalinux

Apr 21 23:52:14 beaglebone pmcd[6570]: Waiting for pmcd to terminate ...
Apr 21 23:52:15 beaglebone pmcd[6570]: Starting pmcd ...
$ sudo systemctl enable pmcd.service
Failed to issue method call: No such file or directory
```

**Comment by kmcdonell** [comment]
OK, so on my Debian 7 system (where all of this just works (tm)), I do NOT have systemd nor systemctl installed.
I just checked here https://wiki.debian.org/systemd/Packaging and looks like integrating the PCP init scripts with Debian systemd is going to take some work.
In the interim, you could try
$ sudo /etc/init.d/pcp start
etc as a workaround?

**Comment by rvernica** [comment]
Got it. Thanks for checking.

Just FYI, I have two identical Debian systems like the one described above. On one of them I went the `./Makepkgs` and `dpkg` route and I have the problem described here. On the second one, I went the `sudo make install` route (see #13) and the `.service` files are installed properly. On the second one, I have a slightly earlier version of pcp but probably it does not matter. I do prefer the `./Makepkgs` route as you suggested.

**Comment by kmcdonell** [comment]
I've done some more investigation on Ubuntu 14.10 (which should be a lot like Rares' Debian system)  and I believe I understand what's going on.

These Debian-based systems have a sort of half-baked systemd set up ... systemd is described as a "technology preview" and we can't commit fully to systemd on this platform because the PCP build is expected to also work for older Debian versions ... so pro tem we're not using systemd on Debian.

And indeed, our debian/rules file contains:
configure_paths = --prefix=/usr --libexecdir=/usr/lib --sysconfdir=/etc --localstatedir=/var --with-rcdir=/etc/init.d --without-systemd --without-python3
(--without-systemd is the key part).

Problem #1 - the PCP init script helper (/usr/share/pcp/lib/rc-proc.sh) are designed to work with all manner of init-rc magic spells ... unfortunately if the systemctl binary is installed it assumes systemctl will work, but this is clearly wrong for the Debian build because the /lib/systemd/system/pm*.service files are missing as a result of the --without-systemd configure option.  I have fixed this and rc-proc.sh is now more careful.  This fixes the initial problem.

Problem #2 - the recommended way to get a configured workarea suitable for running "make; sudo make install" is to first run qa/admin/myconfigure.  This script (a) assumed it would produce the same configuration control files as Makepkgs on the current platform, and (b) did not bother to validate that this did indeed happen for the case when myconfigure is run after the PCP packages have been installed (so you can compare the workarea files and the equivalent installed files, e.g. /etc/pcp.conf, /usr/include/pcp/builddefs, ...).  I have fixed (b) and in the process discovered that (a) was not true for Debian builds.

Problem #3 - the Debian configure magic is replicated in the tree ... in debian/rules and qa/admin/myconfigure and the magic is not the same in both places. I have fixed this, debian/rules is the truth and the only place.  This and the fixes for Problem #2 mean that installing packages made with Makepkgs and "sudo make install" will produce the same results, namely no use of systemd or systemctl or putting files in /lib/systemd/system.

All these changes will flow through my git repo, onto a review process and then to the github repo, so I'm closing this issue in anticipation of this being resolved once this happens.
